### ISSUE TITLE
[Feature Request] SOCKS5, HTTP and TPROXY in one sslocal instance

### ISSUE THREAD

**Issue by database64128** [shadowsocks/shadowsocks-rust#440]
- Add `--listen-socks` as an alias to `-b` and `--local-addr`.
- Add `--listen-http` for HTTP proxy.
- Add `--listen-tproxy` for TPROXY.
- Deprecate and eventually remove `--protocol`.
- In JSON config:
    - Add `listen_socks_address` and `listen_socks_port` as aliases to `local_address` and `local_port`.
    - Add `listen_http_address` and `listen_http_port`.
    - Add `listen_tproxy_address` and `listen_tproxy_port`.
    - Add `forward_address` and `forward_port`.

**Comment by zonyitoo** [comment]
![image](https://user-images.githubusercontent.com/1067951/109962627-63b57e00-7d26-11eb-94bc-d491505952cb.png)

This request includes a lot of changes.

**Comment by zonyitoo** [comment]
Just tried to implement the feature, but I found problems:

1. Should we allow starting multiple locals with the same protocol? For example, `--local-socks-addr 127.0.0.1:1080 --local-socks-addr 127.0.0.1:1081`. Some protocols have specific configurations for a single instance:
    - `redir` requires `--tcp-redir` and `udp-redir`, should all local redir clients shares the same configuration?
    - `tunnel` requires `--forward-addr`, which is clearly shouldn't be shared between these clients
    - `dns` requires `--local-dns-addr` and `--remote-dns-addr`, which are also shouldn't be shared between clients

**Comment by zonyitoo** [comment]
Libev seems to allow binding to multiple addresses. I should support it first.

**Comment by zonyitoo** [comment]
Draft a new configuration for specifying multiple local addresses with different protocols

```jsonc
{
    "locals": [
        {
            "local_address": "127.0.0.1",
            "local_port": 1081
            // socks5 by default
        },
        {
            "local_address": "127.0.0.1",
            "local_port": 3128,
            "protocol": "http"
        },
        {
            "local_address": "0.0.0.0",
            "local_port": 60080,
            "protocol": "redir",
            "tcp_redir": "redirect",
            "udp_redir": "tproxy"
        },
        {
            "local_address": "::",
            "local_port": 60080,
            "protocol": "redir",
            "tcp_redir": "redirect",
            "udp_redir": "tproxy",
            "ipv6_only": true
        },
        {
            "local_address": "::",
            "local_port": 53,
            "protocol": "tunnel",
            "forward_address": "8.8.8.8:53",
            "ipv6_only": false
        }
    ]
}
```

**Comment by database64128** [comment]
Since we don't have routing, multiple listeners of the same protocol aren't really useful in practice.

We don't have to aim for feature parity with ss-libev. Just do things in practical ways.

**Comment by zonyitoo** [comment]
Well, if you are using `redir`, creating 2 listeners which bind to `0.0.0.0` and `::` are very useful for support both IPv4 and IPv6.

**Comment by database64128** [comment]
> Draft a new configuration for specifying multiple local addresses with different protocols

Is it possible to have SOCKS5 remote server support in sslocal?

**Comment by zonyitoo** [comment]
Why would users do that with `sslocal`?

**Comment by database64128** [comment]
> Why would users do that with `sslocal`?

`sslocal` can provide an HTTP proxy.

**Comment by zonyitoo** [comment]
```
HTTP -- sslocal -- SOCKS5 -- sslocal -- ssserver
```

Is this what you mean?

**Comment by zonyitoo** [comment]
If we allows multiple listeners in one `sslocal` instance, what we get are:

1. Easy managing different clients in one configuration file
2. Server statistic informations are shared between different clients

**Comment by database64128** [comment]
More like:

```
  HTTP -+            +- Shadowsocks ----------------------------------------+
        |            |                                                      |
 redir -+-- sslocal -+- SOCKS5 --- V2Ray/Xray/whatever --- other protocols -+- 🌐
        |            |                                                      |
tunnel -+            +- Shadowsocks ----------------------------------------+
```

**Comment by zonyitoo** [comment]
Hah. I don't thik this would be a common use case for shadowsocks users.

But it should be very easy to implement.

**Comment by database64128** [comment]
Personally I only use Shadowsocks servers, so this doesn't really matter to me. But most users have servers from different sources with mixed protocols. shadowsocks-rust has the best ping balancer, and users might want to utilize it for servers of other protocols.

**Comment by zonyitoo** [comment]
Well, supporting SOCKS5 servers won't introduce extra dependencies of libraries, which should be Ok.

**Comment by database64128** [comment]
> Easy managing different clients in one configuration file

If only we can map a certain listener to a certain server.

Currently I have to run 6 instances of shadowsocks-rust on my infrastructure. If we can have listener-server mappings, I can reduce that number down to 4. But that sounds pretty difficult to implement.

**Comment by zonyitoo** [comment]
A simplified version of v2ray.

**Comment by Tigerfyj** [comment]
For the `tunnel` flag, I think one instance can support multiple tunnels if the flag contains the local port as well, like `<local-port>:<remote-address>:<remove-port>`. 
For my use case, I would like to have that as:
```
sslocal ... --udp_tunnel "53:8.8.8.8:53" --udp_tunnel "5353:1.1.1.1:53"
```

**Comment by zonyitoo** [comment]
There would be no problem if we have `locals`: https://github.com/shadowsocks/shadowsocks-rust/issues/440#issuecomment-792004622

**Comment by zonyitoo** [comment]
Merged into master and will be released in v1.10.0.
